### PR TITLE
`constrained-generators`: Hotfix failing test

### PR DESCRIPTION
**Pull request by MaximilianAlgehed**
# Description

It appears to have been an issue with cheating about comparisons and assuming names are unique! They are not for parallel `exists` and we weren't super careful about that.

# Checklist

- [x] Commit sequence broadly makes sense and commits have useful messages
- [x] New tests are added if needed and existing tests are updated
- [x] When applicable, versions are updated in `.cabal` and `CHANGELOG.md` files according to the
      [versioning process](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process).
- [x] The version bounds in `.cabal` files for all affected packages are updated. **If you change the bounds in a cabal file, that package itself must have a version increase.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#versioning-process))
- [x] All visible changes are prepended to the latest section of a `CHANGELOG.md` for the affected packages. **New section is never added with the code changes.** (See [RELEASING.md](https://github.com/intersectmbo/cardano-ledger/blob/master/RELEASING.md#changelogmd))
- [x] Code is formatted with [`fourmolu`](https://github.com/fourmolu/fourmolu) (use `scripts/fourmolize.sh`)
- [x] Cabal files are formatted (use `scripts/cabal-format.sh`)
- [x] [`hie.yaml`](https://github.com/intersectmbo/cardano-ledger/blob/master/hie.yaml) has been updated (use `scripts/gen-hie.sh`)
- [x] Self-reviewed the diff

### CODE REVIEW

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 1]
```diff
@@ -711,7 +722,13 @@ shrinkPred (Block ps) = TruePred : FalsePred [] : ps ++ map Block (shrinkList sh
...
 shrinkPred (Explain _ p) = TruePred : FalsePred [] : [p]
 shrinkPred (When b p) = TruePred : FalsePred [] : p : map (When b) (shrinkPred p)
-shrinkPred (Exists k (x :-> p)) = TruePred : FalsePred [] : [Exists k (x :-> p') | p' <- shrinkPred p]
+-- NOTE: You can't shrink `Exists` because it might make the `Env -> a` invalid!
+-- e.g. you start with
+-- `constrained $ \ x -> exists (\eval -> pure $ eval x) $ \ y -> [ x ==. y, 10 <. y ]`
+-- and shrink it to
+-- `constrained $ \ x -> exists (\eval -> pure $ eval x) $ \ y -> [ 10 <. y ]`
+-- and suddenly the thing you're generating is BS w.r.t the checking!
+shrinkPred Exists {} = [TruePred, FalsePred []]
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Base.hs:731]: Do you mean that the existentially quantified  variable "x" has disappeared?

File: libs/constrained-generators/src/Constrained/Base.hs [hunk 2]
```diff
@@ -4592,6 +4609,7 @@ bind body = x :-> p
 
     bound (Explain _ p) = bound p
     bound (Subst x _ p) = max (nameOf x) (bound p)
+    bound (Block []) = -1
```
Comment (TimSheard) [inline libs/constrained-generators/src/Constrained/Base.hs]: What is going on here?  A comment might help.

**Review by TimSheard — APPROVED** [review]
We used to used an embedded TypeRep to distinguish variables with the same name.
Now we don't, and in the case of existenitals that came back to bite us.

File: libs/constrained-generators/src/Constrained/Base.hs (hunk 1 above)
Comment (MaximilianAlgehed) [reply libs/constrained-generators/src/Constrained/Base.hs:731]: In the example the equality constraint has disappeared

**Pull request merged**